pyasice/xmlsig.py

```diff
@@ -0,0 +1,509 @@
+                raise ValueError("Unsupported signature algorithm")
+        else:
+            algo = self.SIGNATURE_ALGORITHMS[0]
+        sig_method_node = self._get_node("ds:SignatureMethod")
+        sig_method_node.attrib["Algorithm"] = self.SIGNATURE_ALGO_ID_TEMPLATE.format(algo=algo)
+        return self
+
+    def set_root_ca_cert(self, root_cert: Union[Certificate, bytes]) -> "XmlSignature":
+        """Sets a root CA cert. This is not mandatory
+
+        :param root_cert: can be a PEM- or DER-encoded bytes content, or an `oscrypto.Certificate` object
+        """
+        certs_node = self._get_node("xades:CertificateValues")
+        ca_node = etree.Element("{%s}EncapsulatedX509Certificate" % self.NAMESPACES["xades"])
+        ca_node.attrib["Id"] = "%s-ROOT-CA-CERT" % self.NEW_SIGNATURE_ID
+        if not isinstance(root_cert, Certificate):
+            root_cert = load_certificate(root_cert)
+        ca_node.text = base64.b64encode(root_cert.asn1.dump())
+        certs_node.append(ca_node)
+        return self
+
+    def set_ocsp_response(self, ocsp_response: OCSP, embed_ocsp_certificate=False) -> "XmlSignature":
+        """
+        Embed the OCSP response and certificates
+
+        :param OCSP ocsp_response:
+        :param bool embed_ocsp_certificate: Whether to add ocsp certificate to the xml.
+            This is needed when the OCSP service in use
+            does not embed the certificate in its response.
+        :return: self
+        """
+        ocsp_response_node = self._get_node("xades:EncapsulatedOCSPValue")
+        ocsp_response_node.text = base64.b64encode(ocsp_response.get_encapsulated_response())
+
+        if embed_ocsp_certificate:
+            ocsp_certs_node = self._get_node("xades:CertificateValues")
+            ocsp_certs = ocsp_response.get_responder_certs()
+            cert_node = ocsp_certs_node.find(".//xades:EncapsulatedX509Certificate", namespaces=self.NAMESPACES)
+            cert_node.text = base64.b64encode(ocsp_certs[0].dump())
+            cert_node.attrib["Id"] = "S1-Responder-cert-1"
+
+            for i, next_cert in enumerate(ocsp_certs[1:]):
+                cert_node = copy.deepcopy(cert_node)
+                cert_node.text = base64.b64encode(ocsp_certs[next_cert].dump())
+                cert_node.attrib["Id"] = "S1-Responder-cert-%d" % i
+                ocsp_certs_node.append(cert_node)
+
+        return self
+
+    def get_ocsp_response(self) -> Optional[OCSP]:
+        ocsp_response_node = self._get_node("xades:EncapsulatedOCSPValue")
+        try:
+            text = ocsp_response_node.text.strip()
+        except AttributeError:
+            return None
+
+        return OCSP.load(base64.b64decode(text)) if text else None
+
+    def verify_ocsp_response(self) -> "XmlSignature":
+        """Verify embedded OCSP response.
+
+        :raises exceptions.SignatureVerificationError:
+        """
+        self.get_ocsp_response().verify()
```

Comment:
Possibly `None.verify()`, may want `SignatureVerificationError` here instead of `AttributeError`.